### ISSUE TITLE
More GenericArguments

### ISSUE THREAD

**Issue by ST-DDT** [SpongePowered/SpongeAPI#1651]
### The issue

In quite a few commands there are some constants that would also be also valid values in the command element but don't match the format of the data to be parsed/read.

**Example**:
`/settime <noon/dawn/time>`

Currently you have to do something like this:

````java
CommandSpec timeCommandSpec= CommandSpec.builder()
        .arguments(
                GenericArguments.firstParsing(GenericArguments.enum("timeOfDay", TimesOfDay.class), new LocalTimeCommandElement("time")))

        .executor((src, args) -> {
                Optional<TimesOfDay> timeOfDay = args.<LocalTime>getOne("timeOfDay");
                LocalTime time = timeOfDay.map(TimesOfDay::toLocalTime).orElseGet(() -> args.<LocalTime >getOne("time").get());
                world.sendTime(time);
                return CommandResult.success();
            }
        })
````

For this case it is obvious that TimesOfDay is just a string representation of certain constants.

### CommandElementWrapper with constants

Thats why it would be a nice addition to have a `GenericArguments#withConstant(CommandElement, Map<String,?>) : CommandElement` method that could be used to inject those constants.

With that change the code can be simplyified to this:

````java
CommandSpec timeCommandSpec= CommandSpec.builder()
        .arguments(
                GenericArguments.withConstants(new LocalTimeCommandElement("time"), TimesOfDay.getTimeMap()))

        .executor((src, args) -> {
                LocalTime time = args.<LocalTime>getOne("time").get();
                world.sendTime(time);
                return CommandResult.success();
            }
        })
````

### Transforming CommandElements

This is something I considered as an alternative:

`CommandElement#transform(Function<?,?>) : CommandElement`
or
`GenericArguments#transform(CommandElement, Function<?,?>) : CommandElement`

````java
GenericArguments.firstParsing(
    GenericArguments.enum("time", TimesOfDay.class).transform(TimesOfDay::toLocalTime), 
    new LocalTimeCommandElement("time"))
````

which just uses the result of the original `CommandElement` and applies the given function to it.

However I'm not sure whether this would provide good usage results.

### TLDR

* Add `GenericArguments#withConstants(CommandElement, Map<String, ?>) : CommandElement` 
* Add `CommandElement#transform(Function<?,?>) : CommandElement`
or `GenericArguments#transform(CommandElement, Function<?,?>) : CommandElement`

If this gets approved, I can also provide the implementation, but I would like to hear your suggestions first.

**Comment by dualspiral** [comment]
This is supersceded by API-8 stuff, and I think most of what you request is actually possible, particularly with value parameter modifiers. 

I will provide a little caution - in API-8 you really should do as little as possible during parse time because you want parsing to be as quick as possible and only do work if you need to. The notion of "transforming command elements" is really an execute time thing.
